src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java

```diff
@@ -0,0 +1,277 @@
+      InstanceCollection targetInstanceCollection = storage.getInstanceCollection(targetTenantContext);
+
+      Context sourceTenantContext = EventHandlingUtil.constructContext(sharingInstance.getSourceTenantId(),
+        kafkaHeaders.get(OKAPI_TOKEN_HEADER), kafkaHeaders.get(OKAPI_URL_HEADER));
+      InstanceCollection sourceInstanceCollection = storage.getInstanceCollection(sourceTenantContext);
+
+      LOGGER.info("handle :: checking is instance {} exists on tenant {}", instanceId, sharingInstance.getTargetTenantId());
+      getInstanceById(instanceId, sharingInstance.getTargetTenantId(), targetInstanceCollection)
+        .onFailure(failure -> {
+          if (failure.getClass().equals(NotFoundException.class)) {
+            LOGGER.info("handle :: instance {} not found on target tenant: {}", instanceId, sharingInstance.getTargetTenantId());
+            getInstanceById(instanceId, sharingInstance.getSourceTenantId(), sourceInstanceCollection)
+              .onSuccess(srcInstance -> {
+                if (srcInstance == null) {
+                  String errorMessage = format("instance %s not found on source tenant: %s", instanceId, sharingInstance.getSourceTenantId());
+                  sendCompleteEventToKafka(tenantId, sharingInstance, kafkaConfig, record.headers());
+                  promise.fail(errorMessage);
+                } else {
+                  LOGGER.info("handle :: Publishing instance {} with source {} from {} tenant to {} tenant", instanceId, srcInstance.getSource(), sharingInstance.getSourceTenantId(), sharingInstance.getTargetTenantId());
+                  if ("FOLIO".equals(srcInstance.getSource())) {
+                    publishInstanceToTenant(srcInstance, targetInstanceCollection)
+                      .onSuccess(publishedInstance -> {
+                        LOGGER.info("handle :: Updating source to 'CONSORTIUM-FOLIO' for Instance {}", instanceId);
+                        JsonObject jsonInstanceToPublish = srcInstance.getJsonForStorage();
+                        jsonInstanceToPublish.put("source", "CONSORTIUM-FOLIO");
+                        updateInstanceInStorage(Instance.fromJson(jsonInstanceToPublish), sourceInstanceCollection)
+                          .onSuccess(updatesSourceInstance -> {
+                            LOGGER.info("handle :: Source 'CONSORTIUM-FOLIO' updated for Instance {}", instanceId);
+                            sendCompleteEventToKafka(tenantId, sharingInstance, kafkaConfig, record.headers());
+                            promise.complete();
+                          }).onFailure(error -> {
+                            String errorMessage = format("Error update Instance by id %s on the source tenant %s. Error: %s", instanceId, sharingInstance.getTargetTenantId(), error.getCause());
+                            sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
```

Comment:
Where is the error response sent to?

Reply:
Re-reading the code, it seems this is the negative/failure version of sendCompleteEventToKafka. I think sendCompleteEventToKafka should be setup such that it can send positive or negative.